File: src/benchmark_pretty.zig
const std = @import("std");
const pp = @import("zoot").PrettyGoodMachine;

const BenchResult = struct {
    name: []const u8,
    doc_time_ns: u64,
    pick_time_ns: u64,
    completions: usize,
    memo_hits: usize,
    memo_misses: usize,
    memo_entries: usize,
    frontier_size: usize,
    queue_peak: usize,
    overflow: u32,
    height: u32,
};

fn printResult(writer: anytype, result: BenchResult) !void {
    try writer.print("\n=== {s} ===\n", .{result.name});
    try writer.print("  Document construction: {d} µs\n", .{result.doc_time_ns / 1000});
    try writer.print("  Layout search (pick):  {d} µs\n", .{result.pick_time_ns / 1000});
    try writer.print("  Total:                 {d} µs\n", .{(result.doc_time_ns + result.pick_time_ns) / 1000});
    try writer.print("  Layouts: completions={d} frontier={d} peak={d}\n", .{
        result.completions,
        result.frontier_size,
        result.queue_peak,
    });
    try writer.print("  Memo: hits={d} misses={d} entries={d} hit_rate={d:.1}%\n", .{
        result.memo_hits,
        result.memo_misses,
        result.memo_entries,
        if (result.memo_hits + result.memo_misses > 0)
            (@as(f64, @floatFromInt(result.memo_hits)) * 100.0 / @as(f64, @floatFromInt(result.memo_hits + result.memo_misses)))
        else
            0.0,
    });
    try writer.print("  Result: overflow={d} height={d}\n", .{ result.overflow, result.height });
}

// Benchmark 1: Deep concatenation chain (from paper Section 8.1)
fn benchConcatChain(allocator: std.mem.Allocator, n: usize, width: u16) !BenchResult {
    var t = pp.Tree.init(allocator);
    defer t.deinit();

    var timer = try std.time.Timer.start();

    // Build: text("A") ++ text("B") ++ ... (n times)
    var doc = try t.text("A");
    var i: usize = 1;
    while (i < n) : (i += 1) {
        doc = try t.plus(doc, try t.text("B"));
    }

    const doc_time = timer.lap();

    const cost_factory = pp.F2.init(width);
    const best = try t.pick(allocator, cost_factory, doc);
    const pick_time = timer.lap();

    const rank = best.idea.gist.rank;

    return BenchResult{
        .name = "Concat Chain",
        .doc_time_ns = doc_time,
        .pick_time_ns = pick_time,
        .completions = best.stat.completions,
        .memo_hits = best.stat.memo_hits,
        .memo_misses = best.stat.memo_misses,
        .memo_entries = best.stat.memo_entries,
        .frontier_size = best.stat.size,
        .queue_peak = best.stat.peak,
        .overflow = rank.o,
        .height = rank.h,
    };
}

// Benchmark 2: Binary tree S-expression (from paper Section 8.1)
fn benchSExpTree(allocator: std.mem.Allocator, depth: usize, width: u16) !BenchResult {
    var t = pp.Tree.init(allocator);
    defer t.deinit();

    var timer = try std.time.Timer.start();

    const doc = try buildSExpTree(&t, depth);
    const doc_time = timer.lap();

    const cost_factory = pp.F2.init(width);
    const best = try t.pick(allocator, cost_factory, doc);
    const pick_time = timer.lap();

    const rank = best.idea.gist.rank;

    return BenchResult{
        .name = "S-Exp Tree",
        .doc_time_ns = doc_time,
        .pick_time_ns = pick_time,
        .completions = best.stat.completions,
        .memo_hits = best.stat.memo_hits,
        .memo_misses = best.stat.memo_misses,
        .memo_entries = best.stat.memo_entries,
        .frontier_size = best.stat.size,
        .queue_peak = best.stat.peak,
        .overflow = rank.o,
        .height = rank.h,
    };
}

fn buildSExpTree(t: *pp.Tree, depth: usize) !pp.Node {
    if (depth == 0) {
        return try t.text("leaf");
    }

    const left = try buildSExpTree(t, depth - 1);
    const right = try buildSExpTree(t, depth - 1);

    // Build: (lparen, "node", space, left, space, right, rparen)
    var doc = try t.text("(");
    doc = try t.plus(doc, try t.text("node"));
    doc = try t.plus(doc, try t.text(" "));
    doc = try t.plus(doc, left);
    doc = try t.plus(doc, try t.text(" "));
    doc = try t.plus(doc, right);
    doc = try t.plus(doc, try t.text(")"));

    return doc;
}

// Benchmark 3: Many choices (fork-heavy)
fn benchManyChoices(allocator: std.mem.Allocator, n: usize, width: u16) !BenchResult {
    var t = pp.Tree.init(allocator);
    defer t.deinit();

    var timer = try std.time.Timer.start();

    // Build n choices between inline and multiline layouts
    var doc = try t.text("start");
    var i: usize = 0;
    while (i < n) : (i += 1) {
        const inline_opt = try t.plus(
            try t.text(" item"),
            try t.text("X"),
        );
        const multiline_opt = try t.plus(
            try t.plus(pp.Node.nl, try t.text("item")),
            try t.text("X"),
        );
        const choice = try t.fork(inline_opt, multiline_opt);
        doc = try t.plus(doc, choice);
    }

    const doc_time = timer.lap();

    const cost_factory = pp.F2.init(width);
    const best = try t.pick(allocator, cost_factory, doc);
    const pick_time = timer.lap();

    const rank = best.idea.gist.rank;

    return BenchResult{
        .name = "Many Choices",
        .doc_time_ns = doc_time,
        .pick_time_ns = pick_time,
        .completions = best.stat.completions,
        .memo_hits = best.stat.memo_hits,
        .memo_misses = best.stat.memo_misses,
        .memo_entries = best.stat.memo_entries,
        .frontier_size = best.stat.size,
        .queue_peak = best.stat.peak,
        .overflow = rank.o,
        .height = rank.h,
    };
}

// Benchmark 4: Nested indentation
fn benchNestedIndent(allocator: std.mem.Allocator, depth: usize, width: u16) !BenchResult {
    var t = pp.Tree.init(allocator);
    defer t.deinit();

    var timer = try std.time.Timer.start();

    const doc = try buildNestedIndent(&t, depth);
    const doc_time = timer.lap();

    const cost_factory = pp.F2.init(width);
    const best = try t.pick(allocator, cost_factory, doc);
    const pick_time = timer.lap();

    const rank = best.idea.gist.rank;

    return BenchResult{
        .name = "Nested Indent",
        .doc_time_ns = doc_time,
        .pick_time_ns = pick_time,
        .completions = best.stat.completions,
        .memo_hits = best.stat.memo_hits,
        .memo_misses = best.stat.memo_misses,
        .memo_entries = best.stat.memo_entries,
        .frontier_size = best.stat.size,
        .queue_peak = best.stat.peak,
        .overflow = rank.o,
        .height = rank.h,
    };
}

fn buildNestedIndent(t: *pp.Tree, depth: usize) !pp.Node {
    if (depth == 0) {
        return try t.text("body");
    }

    const inner = try buildNestedIndent(t, depth - 1);

    // Build: "{\n" ++ nest(2, inner) ++ "\n}"
    var doc = try t.text("{");
    doc = try t.plus(doc, pp.Node.nl);
    doc = try t.plus(doc, try t.nest(2, inner));
    doc = try t.plus(doc, pp.Node.nl);
    doc = try t.plus(doc, try t.text("}"));

    return doc;
}

// Benchmark 5: JSON-like structure
fn benchJSON(allocator: std.mem.Allocator, n_fields: usize, width: u16) !BenchResult {
    var t = pp.Tree.init(allocator);
    defer t.deinit();

    var timer = try std.time.Timer.start();

    var doc = try t.text("{");
    doc = try t.plus(doc, pp.Node.nl);

    var i: usize = 0;
    while (i < n_fields) : (i += 1) {
        var field = try t.text("  \"field\": \"value\"");

        if (i < n_fields - 1) {
            field = try t.plus(field, try t.text(","));
        }

        field = try t.plus(field, pp.Node.nl);
        doc = try t.plus(doc, field);
    }

    doc = try t.plus(doc, try t.text("}"));

    const doc_time = timer.lap();

    const cost_factory = pp.F2.init(width);
    const best = try t.pick(allocator, cost_factory, doc);
    const pick_time = timer.lap();

    const rank = best.idea.gist.rank;

    return BenchResult{
        .name = "JSON Object",
        .doc_time_ns = doc_time,
        .pick_time_ns = pick_time,
        .completions = best.stat.completions,
        .memo_hits = best.stat.memo_hits,
        .memo_misses = best.stat.memo_misses,
        .memo_entries = best.stat.memo_entries,
        .frontier_size = best.stat.size,
        .queue_peak = best.stat.peak,
        .overflow = rank.o,
        .height = rank.h,
    };
}

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();

    var buffer: [8192]u8 = undefined;
    var stdout_file = std.fs.File.stdout();
    var stdout_writer = stdout_file.writer(&buffer);
    const writer = &stdout_writer.interface;
    defer writer.flush() catch {};

    try writer.writeAll("Pretty Printer Benchmarks (Zig)\n");
    try writer.writeAll("================================\n");
    try writer.print("Build mode: {s}\n", .{@tagName(@import("builtin").mode)});
    try writer.writeAll("\n");

    const width: u16 = 80;

    // Run benchmarks (small sizes for speed)
    {
        const result = try benchConcatChain(allocator, 50, width);
        try printResult(writer, result);
    }

    {
        const result = try benchSExpTree(allocator, 5, width);
        try printResult(writer, result);
    }

    {
        const result = try benchManyChoices(allocator, 10, width);
        try printResult(writer, result);
    }

    {
        const result = try benchNestedIndent(allocator, 5, width);
        try printResult(writer, result);
    }

    {
        const result = try benchJSON(allocator, 20, width);
        try printResult(writer, result);
    }

    try writer.writeAll("\nAll benchmarks completed!\n");
}

File: benchmark_pretty.ml
(* Benchmarks for Pretty_expressive OCaml library *)

module P = Pretty_expressive.Printer.Make(
  val Pretty_expressive.Printer.default_cost_factory ~page_width:80 ()
)

open P

type bench_result = {
  name : string;
  total_time_ns : int64;
  doc_size : int; (* approximate indicator *)
}

let time_ns f =
  let start = Unix.gettimeofday () in
  let result = f () in
  let finish = Unix.gettimeofday () in
  let time_ns = Int64.of_float ((finish -. start) *. 1e9) in
  (result, time_ns)

let print_result result =
  Printf.printf "\n=== %s ===\n" result.name;
  Printf.printf "  Total time: %Ld µs\n" (Int64.div result.total_time_ns 1000L);
  Printf.printf "  Doc size indicator: %d\n" result.doc_size

(* Benchmark 1: Deep concatenation chain *)
let bench_concat_chain n =
  let (output, time_ns) = time_ns (fun () ->
    let rec build_chain i doc =
      if i >= n then doc
      else build_chain (i + 1) (doc ^^ text "B")
    in
    let doc = build_chain 1 (text "A") in
    pretty_format doc
  ) in
  {
    name = "Concat Chain";
    total_time_ns = time_ns;
    doc_size = String.length output;
  }

(* Benchmark 2: Binary tree S-expression *)
let rec build_sexp_tree depth =
  if depth = 0 then
    text "leaf"
  else
    let left = build_sexp_tree (depth - 1) in
    let right = build_sexp_tree (depth - 1) in
    lparen ^^ text "node" ^^ space ^^ left ^^ space ^^ right ^^ rparen

let bench_sexp_tree depth =
  let (output, time_ns) = time_ns (fun () ->
    let doc = build_sexp_tree depth in
    pretty_format doc
  ) in
  {
    name = "S-Exp Tree";
    total_time_ns = time_ns;
    doc_size = String.length output;
  }

(* Benchmark 3: Many choices *)
let bench_many_choices n =
  let (output, time_ns) = time_ns (fun () ->
    let rec build_choices i doc =
      if i >= n then doc
      else
        let inline_opt = space ^^ text "item" ^^ text (string_of_int i) in
        let multiline_opt = nl ^^ text "item" ^^ text (string_of_int i) in
        let choice = inline_opt <|> multiline_opt in
        build_choices (i + 1) (doc ^^ choice)
    in
    let doc = build_choices 0 (text "start") in
    pretty_format doc
  ) in
  {
    name = "Many Choices";
    total_time_ns = time_ns;
    doc_size = String.length output;
  }

(* Benchmark 4: Nested indentation *)
let rec build_nested_indent depth =
  if depth = 0 then
    text "body"
  else
    let inner = build_nested_indent (depth - 1) in
    lbrace ^^ nl ^^ nest 2 inner ^^ nl ^^ rbrace

let bench_nested_indent depth =
  let (output, time_ns) = time_ns (fun () ->
    let doc = build_nested_indent depth in
    pretty_format doc
  ) in
  {
    name = "Nested Indent";
    total_time_ns = time_ns;
    doc_size = String.length output;
  }

(* Benchmark 5: JSON-like structure *)
let bench_json n_fields =
  let (output, time_ns) = time_ns (fun () ->
    let rec build_fields i doc =
      if i >= n_fields then doc
      else
        let key = Printf.sprintf "field%d" i in
        let value = Printf.sprintf "value%d" i in
        let field = text "  " ^^ dquote ^^ text key ^^ text "\": \"" ^^
                    text value ^^ dquote in
        let field = if i < n_fields - 1 then field ^^ comma else field in
        let field = field ^^ nl in
        build_fields (i + 1) (doc ^^ field)
    in
    let doc = lbrace ^^ nl ^^ build_fields 0 empty ^^ rbrace in
    pretty_format doc
  ) in
  {
    name = "JSON Object";
    total_time_ns = time_ns;
    doc_size = String.length output;
  }

let () =
  print_endline "Pretty Printer Benchmarks (OCaml)";
  print_endline "==================================";
  print_endline "";

  (* Run benchmarks with same parameters as Zig *)
  let r1 = bench_concat_chain 50 in
  print_result r1;

  let r2 = bench_sexp_tree 5 in
  print_result r2;

  let r3 = bench_many_choices 10 in
  print_result r3;

  let r4 = bench_nested_indent 5 in
  print_result r4;

  let r5 = bench_json 20 in
  print_result r5;

  print_endline "";
  print_endline "All benchmarks completed!"
